api/src/pcapi/routes/backoffice/bookings/individual_bookings_blueprint.py

```diff
@@ -404,52 +317,161 @@ def get_batch_cancel_individual_bookings_form() -> utils.BackofficeResponse:
+        with repository.atomic():
+            token = booking.token
+            try:
+                bookings_api.mark_as_cancelled(
+                    booking=booking,
+                    reason=reason,
+                    author_id=current_user.id,
+                )
+                success_count += 1
+            except (
+                bookings_exceptions.BookingIsAlreadyUsed,
+                bookings_exceptions.BookingIsAlreadyRefunded,
+                bookings_exceptions.BookingIsAlreadyCancelled,
+            ) as exc:
+                error_dict[exc.__class__.__name__].append(token)
+                repository.mark_transaction_as_invalid()
+            except (
+                cgr_exceptions.CGRAPIException,
+                cds_exceptions.CineDigitalServiceAPIException,
+                ems.EMSAPIException,
+            ) as exc:
+                logger.info(
+                    "API error for cancelling external booking, the booking will be cancelled unilaterally",
+                    extra={"booking_id": booking.id, "exc": str(exc)},
+                )
+                try:
+                    bookings_api.mark_as_cancelled(
+                        booking=booking,
+                        reason=reason,
+                        one_side_cancellation=True,
+                        author_id=current_user.id,
+                    )
+                    success_count += 1
+                except Exception as exception:  # pylint: disable=broad-except
+                    repository.mark_transaction_as_invalid()
+                    flash(Markup("Une erreur s'est produite : {message}").format(message=str(exception)), "warning")
+            except Exception as exc:  # pylint: disable=broad-except
+                repository.mark_transaction_as_invalid()
+                flash(Markup("Une erreur s'est produite : {message}").format(message=str(exc)), "warning")
+
+    _flash_success_and_error_message(success_count, error_dict, False)
+
+
+def _flash_success_and_error_message(success_count: int, error_dict: dict[str, list[str]], is_validating: bool) -> None:
+    if success_count > 0:
+        flash(
+            f"{success_count} {pluralize(success_count, 'réservation a', 'réservations ont')} été {'validée'+pluralize(success_count) if is_validating else 'annulée'+pluralize(success_count)}",
+            "success",
+        )
 
-    return _redirect_after_individual_booking_action()
+    if error_dict:
+        if success_count > 0:
+            error_text = f"""Certaines réservations n'ont pas pu être {'validées' if is_validating else 'annulées'} et ont été ignorées :<br> """
+        else:
+            error_text = f"""Impossible {"de valider" if is_validating else "d'annuler"} ces réservations : <br>"""
+        for key in error_dict:
+            match key:
+                case "BookingIsAlreadyCancelled":
+                    error_text += _build_booking_error_str(
+                        error_dict[key],
```

Comment:
`tokens = error_dict[key]` juste sous le `for` permettrait de rendre le code plus digeste en accédant juste à la variable dans tous les nombreux appels.